bloom.go

```diff
@@ -0,0 +1,89 @@
+// Package bloom implements a Bloom filter data structure.
+package bloom
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"log"
+	"math/bits"
+)
+
+// Filter represents a Bloom filter. Filter satisfies the encoding.BinaryMarshaler and BinaryUnmarshaler interfaces.
+// The zero value represents an empty filter of size 0 that uses 0 hash values.
+// Current implementations of Insert and Contains require len(f) <= 2^16 bits and k <= 16.
+type Filter struct {
+	f []byte
+	k int
+}
+
+// bit returns the filter's nth bit.
+func (f *Filter) bit(n int) int {
+	b, i := n/8, n%8
+	return int(f.f[b]>>uint(i)) & 1
+}
+
+// setBit sets the filter's nth bit to 1.
+func (f *Filter) setBit(n int) {
+	b, i := n/8, n%8
+	f.f[b] |= 1 << uint(i)
+}
+
+// hashBits returns the base 2 logarithm of the size of the filter in bits.
+// This is the number of hash bits required to index any bit in the filter.
+// TODO: Handle filter size not a power of 2
+func (f *Filter) hashBits() int {
+	if bits.OnesCount(uint(len(f.f))) != 1 {
+		log.Fatal("hashBits: filter size not a power of 2")
+	}
+	return bits.TrailingZeros64(uint64(len(f.f)) * 8)
+}
+
+// New returns a Filter of size b bytes that uses k hash values.
+func New(b, k int) *Filter { return &Filter{make([]byte, b), k} }
+
+// Insert inserts data into f's set.
+func (f *Filter) Insert(data []byte) {
+	hash := sha256.Sum256(data)
+	for h := 0; h < f.k; h++ {
+		// i is constructed from two bytes, limiting filter size to 2^16 bits.
+		// The factor of 2 also limits f.k to 16 hash values.
+		i := binary.BigEndian.Uint16(hash[2*h:]) & (1<<uint(f.hashBits()) - 1)
+		f.setBit(int(i))
+	}
+}
+
+// Contains reports whether data is in f's set.
+// If Contains returns true, a false positive is possible,
+// but if Contains returns false, data is definitely not in the set.
+func (f *Filter) Contains(data []byte) bool {
+	hash := sha256.Sum256(data)
+	for h := 0; h < f.k; h++ {
+		// i is constructed from two bytes, limiting filter size to 2^16 bits.
+		// The factor of 2 also limits f.k to 16 hash values.
+		i := binary.BigEndian.Uint16(hash[2*h:]) & (1<<uint(f.hashBits()) - 1)
```

Comment:
It seems like there's an opportunity to factor out some nontrivial & subtle logic that is common to Insert and Contains. Maybe that doesn't work out so nice, but I would try playing around with it if you didn't already.

Reply:
I went back and forth on factoring out the conversion of the [32]byte hash to a [16]uint16 to be iterated over:
`u := uints(hash)
for h := 0; h < f.k; h++ {
	// Operate on f using u[h]
}`
It's not much of an improvement at present, because the mechanism by which the uint16s are composed is fairly simplistic. But it does have the corresponding advantage of abstracting away a single function responsible for dictating how the hash bits are used, if and when a future commit should decide to use them more efficiently, and that's the right thing to do.